File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-3.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        maxArea=0
        stack=[]
        for pos,height in enumerate(heights):
            if len(stack)==0:
                stack.append([pos,height])
                maxArea=height
            elif height<stack[-1][1]:
                while len(stack)>0 and height<stack[-1][1]:
                    oldpos,oldheight=stack.pop()
                    maxArea=max(maxArea,(pos-oldpos)*oldheight)
                stack.append([oldpos,height])
            else:
                stack.append([pos,height])
        length=len(heights)
        for pos,height in stack:
            maxArea=max(maxArea,(length-pos)*height)
        return maxArea

File: Data Structures & Algorithms/max-water-container/submission-3.java
class Solution {
    public int maxArea(int[] heights) {
        int l=0;
        int r = heights.length-1;
        int maxVal=0;
        while (r>l){
            int vol=(r-l)*Math.min(heights[r],heights[l]);
            maxVal=Math.max(vol,maxVal);
            if (heights[r]>=heights[l]){
                l++;
            }else{
                r--;
            }
        }
        return maxVal;
    }
}
